File: #227_Basic_Calculator_II/Solution.java
public class Solution {

	int index;

    public int calculate(String s) {
    	String str = trimSpace(s);
    	if (str == null || str.length() == 0) return 0;
    	index = 0;
    	int length = str.length();
    	long sum = 0;
    	boolean negative = false;
    	while (index < length) {
    		int start = index;
    		while (index < length && isDigit(str.charAt(index))) index++;
    		if (index != start) { // find a number
    			long num = Long.valueOf(str.substring(start, index));
    			if (index == length) return (negative) ? (int) (sum - num) : (int) (sum + num);
    			if (str.charAt(index) == '*' || str.charAt(index) == '/') { // multiplication / division
    				index = start;
    				sum = (negative) ? sum - multdiv(str) : sum + multdiv(str);
    			}
    			else sum = (negative) ? sum - num : sum + num;
    			negative = false;
    			continue;
    		}
    		if (str.charAt(index) == '-') negative = true;
    		index++;
    	}
    	return (int) sum;
    }

    private long multdiv(String string) {
    	if (string == null || string.length() == 0) return 0;
    	long result = 1;
    	boolean div = false;
    	while (true) {
    		int start = index;
    		while (index < string.length() && isDigit(string.charAt(index))) index++;
    		long num = Long.valueOf(string.substring(start, index));
    		result = (div) ? result / num : result * num;
    		div = false;
    		if (index == string.length() || string.charAt(index) == '+' || string.charAt(index) == '-') return result;
    		if (string.charAt(index) == '/') div = true;
			index++;
    	}
    }

    private String trimSpace(String s) {
    	if (s == null || s.length() == 0) return s;
    	StringBuilder sb = new StringBuilder();
    	for (int i = 0; i < s.length(); i++) {
    		if (s.charAt(i) != ' ') sb.append(s.charAt(i));
    	}
    	return sb.toString();
    }

    private boolean isDigit(char c) {
		return c >= '0' && c <= '9';
	}

	public static void main(String[] args) {
		Solution sol = new Solution();
		System.out.println(sol.calculate("14/3*2"));
	}
}

File: #227_Basic_Calculator_II/Solution.py
class Solution(object):
    def calculate(self, s):
        if not s:
        	return 0
        stack = []
        re = 0
        num = 0
        sign = 1 # 1 for +, -1 for -, 2 for *, 3 for /
        for c in s:
        	if c.isdigit():
        		num = num * 10 + ord(c) - ord('0')
        	elif c == '+' or c == '-':
        		if stack:
        			if sign == 2:
        				re *= num
        			if sign == 3:
        				re /= num
        			re *= stack.pop()
        			re += stack.pop()
        		else:			
		    		re += num * sign
	    		num = 0
	    		if c == '+':
	    			sign = 1
	    		else:
	    			sign = -1
        	elif c == '*' or c =='/':
        		if not stack:
	        		stack.append(re)
	        		stack.append(sign)
	        		re = num
	        	else:
	        		if sign == 2:
	        			re *= num
	        		if sign == 3:
	        			re /= num
        		num = 0
        		if c == '*':
        			sign = 2
        		else:
        			sign = 3
        if num != 0:
        	if sign == 2:
        		re *= num
        	elif sign == 3:
        		re /= num
    		if stack:
    			re *= stack.pop()
    			re += stack.pop()
    		if sign == -1 or sign == 1:
    			re += sign * num
        return re

sol = Solution()
print sol.calculate("1 + 2 * 3")